proof/crefine/X64/Interrupt_C.thy

```diff
@@ -573,23 +602,49 @@ lemma ioapic_decode_map_pin_to_vector_ccorres:
                             EXCEPTION_SYSCALL_ERROR_def EXCEPTION_NONE_def syscall_error_rel_def)
       apply (clarsimp simp: rf_sr_def cstate_relation_def carch_state_relation_def Let_def)
       apply (subst ucast_sub_ucast; fastforce simp: lt1_neq0)
+     apply (rule_tac P="numIOAPICs \<le> of_nat maxNumIOAPIC" in ccorres_gen_asm)
+     apply (clarsimp simp: not_less word_le_def[symmetric])
+     apply (prop_tac "ioapic < of_nat maxNumIOAPIC",
+            solves \<open>simp add: le_m1_iff_lt[THEN iffD1] word_neq_0_conv\<close>)
+     apply (simp add: Kernel_Config.maxNumIOAPIC_def)
+     apply ccorres_rewrite
+     apply (rename_tac ioapic_nirqs)
+     apply (rule_tac Q="\<lambda>s. ioapic_nirqs = x64KSIOAPICnIRQs (ksArchState s) \<and>
+                            (0 < x64KSIOAPICnIRQs (ksArchState s) ioapic)" and Q'=\<top>
+                     in ccorres_split_when_throwError_cond)
+        apply (fastforce simp: word_le_def scast_ucast_up_eq_ucast uint_up_ucast is_up
+                               rf_sr_x64KSIOAPICnIRQs Kernel_Config.maxNumIOAPIC_def
+                               uint_less_1_eq)
+      (* Need to VCG it as the range error depends on the global state *)
```

Comment:
Thanks for these comments.

Reply:
I have the feeling I copied this from somewhere and shouldn't take credit for it. It sounds like something Raf might have written. But I approve as well :-)